library/std/src/io/stdio.rs

```diff
@@ -1017,6 +1017,103 @@ where
     }
 }
 
+// This needs a dedicated separate Sealed trait, not just `sealed::Sealed`, because it will have
+// blanket impls for other traits.
+mod terminal_sealed {
+    #[unstable(feature = "sealed", issue = "none")]
+    pub trait Sealed {}
+}
+
+/// Trait to determine if a descriptor/handle refers to a terminal/tty.
+#[unstable(feature = "is_terminal", issue = "80937")]
+pub trait IsTerminal: terminal_sealed::Sealed {
+    /// Returns true if the descriptor/handle refers to a terminal/tty.
+    ///
+    /// Returns an error if any errors occurred when attempting to make this determination, other
+    /// than those expected as part of making this determination. For instance, on some platforms
+    /// checking for a terminal may involve making a system call that returns an error on a
+    /// non-terminal, so that error will get translated to `Ok(false)`, but any unexpected error
+    /// will still get returned as an error.
+    fn is_terminal(&self) -> io::Result<bool>;
+}
+
+cfg_if::cfg_if! {
+    if #[cfg(unix)] {
+        use crate::os::unix::io::{AsFd, AsRawFd};
+
+        #[unstable(feature = "sealed", issue = "none")]
+        impl<F: AsFd> terminal_sealed::Sealed for F {}
+
+        #[unstable(feature = "is_terminal", issue = "80937")]
+        impl<F: AsFd> IsTerminal for F {
+            fn is_terminal(&self) -> io::Result<bool> {
+                let fd = self.as_fd();
+                let ret = unsafe { libc::isatty(fd.as_raw_fd()) };
+                if ret == 0 {
+                    let err = io::Error::last_os_error();
+                    if err.raw_os_error() == Some(libc::ENOTTY) {
+                        return Ok(false);
+                    }
+                    return Err(err);
+                }
+                return Ok(true);
+            }
+        }
+    } else if #[cfg(windows)] {
+        use crate::os::windows::io::AsHandle;
+
+        #[unstable(feature = "sealed", issue = "none")]
+        impl<H: AsHandle> terminal_sealed::Sealed for H {}
+
+        #[unstable(feature = "is_terminal", issue = "80937")]
+        impl<H: AsHandle> IsTerminal for H {
+            fn is_terminal(&self) -> io::Result<bool> {
+                unsafe { crate::sys::io::handle_is_console(self.as_handle()) }
+            }
+        }
+    } else {
+        #[unstable(feature = "sealed", issue = "none")]
+        impl terminal_sealed::Sealed for Stdin {}
+
+        #[unstable(feature = "is_terminal", issue = "80937")]
+        impl IsTerminal for Stdin {
+            fn is_terminal() -> io::Result<bool> {
+                Ok(false)
```

Comment:
Shouldn't platforms which might have terminals but don't fit our unix/windows default implementations return an error with `ErrorKind::Unsupported` instead? I.e. tell the caller that this may or may not be a terminal and we can't tell.

Reply:
We could do that, but then users writing portable software would need to expect such errors and turn them to `false`, rather than aborting. Under what circumstances can you envision anyone wanting to do something other than treat it as not a terminal?

We could return `Result<Option<bool>>`, but I cannot think of a single circumstance in which that would be useful.

Reply:
It just seems odd to have an default-Ok on platforms where it's not implemented and the potential for actual errors on platforms where it's implemented.

Reply:
@the8472 Fair. I do think that it's useful to be able to error out if a call to the system fails. But I think that's orthogonal to the question of what semantic we're offering for cases we can't determine.

I think we should make it possible for people to call `.is_terminal()?`, and not make that systematically fail on some platforms.

Reply:
I've added documentation explaining the behavior on unsupported platforms, since we shouldn't leave that undocumented.